subsys/console/tty.c

```diff
@@ -68,10 +69,18 @@ static int tty_irq_input_hook(struct tty_serial *tty, u8_t c)
 	return 1;
 }
 
-int tty_putchar(struct tty_serial *tty, u8_t c)
+ssize_t tty_write(struct tty_serial *tty, const void *buf, size_t size)
 {
+	u8_t c = *(const u8_t *)buf;
 	unsigned int key;
 	int tx_next;
+	int res;
+
+	res = k_sem_take(&tty->tx_sem, tty->tx_timeout);
```

Comment:
this doesn't look right. There might be room in the ring buffer but thread will get blocked until irq is handled. I would assume that this happens only when ringbuffer gets full. Irq_locking is already protecting the ring buffer.

Reply:
> this doesn't look right. There might be room in the ring buffer but thread will get blocked until irq is handled.

Did you see how sema's are initialized? rx_sema starts with zero, because rx buffer starts as empty. tx_sema start as tx_buf_sz-1, because that's how much free space we have there. If user asked us to block on TX if it's not possible to process (buffer in our case) tx data, we'll do that until ISR finds some free space for us, indeed.

Reply:
ok, didn't noticed that.